### PR TITLE
[SPARK-12792][SPARKR] Refactor RRDD to support R UDF.

### PR DESCRIPTION
**Pull request by sun-rui**
Refactor RRDD by separating the common logic interacting with the R worker to a new class RRunner, which can be used to evaluate R UDFs.

Now RRDD relies on RRuner for RDD computation and RRDD could be reomved if we want to remove RDD API in SparkR later.

### CODE REVIEW

**Comment by shivaram** [comment]
cc @davies. Thanks @sunrui for the PR. I'll review this later today

**Comment by sun-rui** [comment]
This PR happens to solve the concurrent computation issue of RRDD reported in SPARK-13178. Added a test case for it.

**Comment by sun-rui** [comment]
Jenkins, retest this please

**Comment by sun-rui** [comment]
Jenkins, retest this please

**Comment by sun-rui** [comment]
Jenkins, retest this please

**Comment by NarineK** [comment]
Thanks for working on this, @sun-rui .
I have a question regarding the R workers. Do we assume that there is a R instance installed on each cluster node ? Is that being considered in the design ?
Thanks,
Narine

**Comment by sun-rui** [comment]
@NarineK, yes it is assumed that R has been installed in each worker node.

**Comment by sun-rui** [comment]
@davies, @rxin, could you help to review it and merge it? it will benefit development of following PRs.

**Comment by rxin** [comment]
For bulk movement of code, can you comment in the pr yourself which part was actually changed, and which part was simply moving code from one place to another?

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+  def getLines(): String = synchronized {
+    (0 until errBufferSize).filter { x =>
+      lines((x + lineIdx) % errBufferSize) != null
+    }.map { x =>
+      lines((x + lineIdx) % errBufferSize)
+    }.mkString("\n")
+  }
+}
+
+private[r] object RRunner {
```
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: Below code are moved from object RRDD with no change.

**Comment by rxin** [comment]
cc @davies

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+
+  private def readStringData(length: Int): String = {
+    length match {
+      case length if length > 0 =>
+        SerDe.readStringBytes(dataStream, length)
+      case _ => null
+    }
+  }
+}
+
```
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: Below code are moved from RRDD with no change.

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+          read()
+        case length if length >= 0 =>
+          readData(length).asInstanceOf[U]
+      }
+    } catch {
+      case eof: EOFException =>
+        throw new SparkException("R worker exited unexpectedly (cranshed)", eof)
+    }
+  }
+
```
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: Below is newly added code.
Basically this logic are originally implemented in the subclasses of BaseRRDD by overriding the readData() protected method.  readData() is used to read data returned from the R worker. Data returned from the R worker is in different format according to different sub-classes of BaseRRDD.

In RRunner, 
PairwiseRDD.readData() moved to  readShuffledData()
RRDD.readData() moved to  readByteArrayData()
StringRRDD.readData() moved to  readStringData()

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+                boot,
+                init,
+                broadcast,
+                input,
+                compute,
+                output,
+                boot + init + broadcast + input + compute + output))
+          read()
+        case length if length >= 0 =>
+          readData(length).asInstanceOf[U]
```
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: readData is a val holding a function for reading data returned from the R worker.
It is one of the following 3 functions:
readShuffledData(), readByteArrayData(), readStringData()

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+    func: Array[Byte],
+    deserializer: String,
+    serializer: String,
+    packageNames: Array[Byte],
+    broadcastVars: Array[Broadcast[Object]],
+    numPartitions: Int = -1)
+  extends Logging {
+  private var bootTime: Double = _
+  private var dataStream: DataInputStream = _
+  val readData = numPartitions match {
```
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: line 44 ~line 51 are new code.

readData is a val holding a function for reading data returned from the R worker.
It is one of the following 3 functions:
readShuffledData(), readByteArrayData(), readStringData()

According the constructor parameters, we can decide which function is used.

File: core/src/main/scala/org/apache/spark/api/r/RRunner.scala (hunk 4 above)
Comment (sun-rui) [inline core/src/main/scala/org/apache/spark/api/r/RRunner.scala]: call the function in readData to return data from the R worker.

**Comment by rxin** [comment]
hey @davies when you have a chance can you review this? thanks.

**Comment by davies** [comment]
LGTM, merging this into master, thanks!

**Comment by davies** [comment]
@sun-rui This PR have some conflict that fail the build, so reverted it, could you send another PR to fix that?

**Comment by sun-rui** [comment]
ok, will send another PR soon.

**Pull request closed without merging**